Abstract tx management in ExecutionEngine

davidegrohmann opened the pull request.

davidegrohmann committed:
TransactionalContext expose the minimun methods needed in the compiler code

davidegrohmann committed:
Make v2_3.TransactionBoundQueryContext use TransactionBoundTransactionalContext

davidegrohmann committed:
ABstract tx management in ExecugineEngine

davidegrohmann committed:
Cleanups after rebase

pontusmelke commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/spi/QueryContext.scala:
```diff
@@ -208,7 +198,6 @@ trait TransactionalContext[Graph,KernelStatement,StateView] {
-  def graph: Graph
-
-  def stateView: StateView
+  // this should not be here, but it is needed by the code generation
```
Why are we putting this back, we just got rid of this dependency?

davidegrohmann replied:
It is needed since the statement is not exposed anymore by this interface.

pontusmelke replied:
Can't you use the same trick we used before and have

```
type ReadOps
...
def readOperations: ReadOps
```

in `QueryContext` and then in the implementing class (`Neo4jTransactionContext`) you do

```
type ReadOps = org.neo4j.kernel.api.ReadOperations
...
def readOperations = statement.readOperations()
```

?

davidegrohmann replied:
Yeah I think I can. I'll do that.

davidegrohmann committed:
Add generic `ReadOps` type to TransactionalContext

In this way we can remove the dependency to the kernel from the cypher
compiler

Mats-SX commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/Neo4jTransactionContext.scala:
```diff
@@ -17,24 +17,35 @@
+  }
+}
+
+// please construct this class through TransactionContextFactory, this is public only for tests
```
I feel like this comment is bound to be overlooked at some point. Is there another way to design this, and still allow for testing? A test subclass?

davidegrohmann replied:
I can try to clean this up, but I would prefer to do so in a separate PR.

Mats-SX replied:
That's fine.
